cpp/src/arrow/util/bit_util.h

```diff
@@ -827,6 +834,164 @@ Status InvertBitmap(MemoryPool* pool, const uint8_t* bitmap, int64_t offset,
+
+  bool operator[](int64_t i) const { return GetBit(i); }
+
+  void SetBitTo(int64_t i, bool v) const {
+    BitUtil::SetBitTo(buffer_->mutable_data(), i + offset_, v);
+  }
+
+  /// \brief Visit bits of this bitmap either as bitset<N> or as array<Word, N>.
+  ///
+  /// If visiting by words, note that the every Word containing a bit
+  /// must be accessible (and initialized, or Valgrind will barf).
+  /// This is satisfied by most Buffers in Arrow space (which are allocated with
+  /// maximum alignment and padded to contain whole cache lines).
+  ///
+  /// TODO(bkietz) allow for early termination
+  template <size_t N, typename Visitor>
+  static void Visit(const Bitmap (&bitmaps)[N], Visitor&& visitor) {
+    VisitImpl(bitmaps, VisitedIs<internal::call_traits::argument_type<0, Visitor&&>>{},
+              std::forward<Visitor>(visitor));
+  }
+
+  const std::shared_ptr<Buffer>& buffer() const { return buffer_; }
+
+  /// offset of first bit relative to buffer().data()
+  int64_t offset() const { return offset_; }
+
+  /// number of bits in this Bitmap
+  int64_t length() const { return length_; }
+
+  /// string_view of all bytes which contain any bit in this Bitmap
+  util::basic_string_view<uint8_t> bytes() const {
+    auto byte_offset = offset_ / 8;
+    auto byte_count = BitUtil::CeilDiv(offset_ + length_, 8) - byte_offset;
+    return util::basic_string_view<uint8_t>(buffer_->data() + byte_offset, byte_count);
+  }
+
+  template <typename Word>
+  bool word_aligned() const {
+    auto words = this->words<Word>();
+    auto valid_addr = AsInt(buffer_->data());
+    auto words_addr = AsInt(words.data());
+    return words_addr >= valid_addr &&
+           words_addr + words.size() * sizeof(Word) <= valid_addr + buffer_->capacity();
+  }
+
+  /// string_view of all Words which contain any bit in this Bitmap
+  template <typename Word>
+  util::basic_string_view<Word> words() const {
+    auto bytes_addr = AsInt(bytes().data());
+    auto words_addr = bytes_addr - bytes_addr % sizeof(Word);
+    auto word_byte_count =
+        BitUtil::RoundUpToPowerOf2(bytes_addr + bytes().size(), sizeof(Word)) -
+        words_addr;
+    return util::basic_string_view<Word>(reinterpret_cast<const Word*>(words_addr),
+                                         word_byte_count / sizeof(Word));
+  }
+
+  /// offset of first bit relative to words<Word>().data()
+  template <typename Word>
+  int64_t word_offset() const {
+    return offset_ + (AsInt(buffer_->data()) - AsInt(words<Word>().data())) * 8;
+  }
+
+ private:
+  template <typename>
+  struct VisitedIs {};
+
+  static size_t AsInt(const void* ptr) {
+    return reinterpret_cast<size_t>(reinterpret_cast<const char*>(ptr));
+  }
+
+  template <size_t N, typename Visitor>
+  static void VisitImpl(const Bitmap (&bitmaps)[N], VisitedIs<std::bitset<N>>,
+                        Visitor&& visitor) {
+    auto bit_length = bitmaps[0].length();
+    std::bitset<N> bits;
```

Comment:
I definitely think `bitset<N>` will be faster; for one thing it's on the stack (under the hood it's backed by `array<uint64_t, WordsForBits(N)>`, IIRC). Since `N` is constexpr (and usually very small) I also expect the loop to be unrolled in most cases.

Reply:
You could, but `std::bitset` should be equivalent or even more efficient than that